added JAX interop adjoint support using FFI

mehdiataei opened the pull request:
<!--
Thank you for contributing to NVIDIA Warp!

See the contribution guide: https://nvidia.github.io/warp/modules/contribution_guide.html

Please fill the relevant sections.

Checkboxes can also be marked after you submit the PR.
-->

## Description
<!--
-->

Added JAX interop adjoint support using FFI. Also added support for static inputs (with auto-detection), vmap, multi-output with AD. Fixed some bugs related to vmap for the forward pass. 

This is a new area for me, so there may be some oddities or poor decisions, but all tests pass. We may need to discuss any missing features and whether this is the way the user intends to use it.

This work is inspired by and thanks to J. Sevcik
https://gist.github.com/jaro-sevcik/893ffb891564dfc7617c8f12f3c2d1d2

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] Necessary tests have been added
- [x] Documentation is up-to-date
- [x] Auto-generated files modified by compiling Warp and building the documentation have been updated (e.g. `__init__.pyi`, `functions.rst`)
- [x] Code passes formatting and linting checks with `pre-commit run -a`

johnpjf commented on warp/tests/interop/test_jax.py:
```diff
@@ -308,6 +308,533 @@ def f_2d():
+    s = 2.0
+
+    with jax.default_device(wp.device_to_jax(device)):
+        da, db = jax.grad(loss, argnums=(0, 1))(a, b, s)
```
This takes a grad of jit, but it's more typical to take the jit of grad. Will this work, can we add a test for it?

mehdiataei replied:
I'll add a test for it

johnpjf reviewed (changes requested):
Thanks for doing this!
I left one comment, but another overall comment is whether we can get a test of this working with multiple devices? I think the older way to do this would be within a `jax.pmap` but there may be newer ways.

johnpjf commented:
FYI, I'm trying this on a two GPU machine and getting an error:
```

@wp.kernel
def triple_kernel(input: wp.array(dtype=float), output: wp.array(dtype=float)):
  tid = wp.tid()  # pytype: disable=module-attr
  output[tid] = 3.0 * input[tid]

class WarpDeferredTest(googletest.TestCase):

  def test_give_me_a_name(self):

    jax_triple = jax_experimental.jax_ad_kernel(triple_kernel)

    def f(x):
      return jax_triple(x)

    @jax.pmap
    def compute_grads(x):
      def loss_fn(x):
        return jnp.sum(jnp.square(f(x)[0]))

      loss, grads = jax.value_and_grad(
          loss_fn, has_aux=False)(x)
      return grads

    x = jnp.broadcast_to(jnp.arange(4, dtype=jnp.float32)[jnp.newaxis], (jax.local_device_count(), 4))
    print(compute_grads(x))



if __name__ == "__main__":
  googletest.main()

```

This fails with:

```
Module __main__ 89dc8ef load on device 'cuda:0' took 318.63 ms  (compiled)
Warp CUDA error 400: invalid resource handle (in function wp_cuda_launch_kernel, third_party/py/warp/native/warp.cu:4061)
```

Notably, if I hide one of the GPUs (with `export CUDA_VISIBLE_DEVICES=0`) so that I only have 1, the tests passes.

mehdiataei commented:
Hey @johnpjf 
Yes, if the examples are not executed with mpirun, we get that error. I’m not sure what the root cause is, but it’s probably a stream, event, function, or module created on GPU 0 being accidentally used while GPU 1 is current. Running with mpirun fixes it. I made the following test cases to check whether the grad works with shardmap, and they all pass successfully. However, I’m having trouble merging these tests with the current Warp test cases since they rely on mpirun.

`import warp as wp
import jax
import jax.numpy as jnp
import numpy as np
from jax.sharding import PartitionSpec as P
from jax.experimental.multihost_utils import process_allgather as allgather
from jax.experimental.shard_map import shard_map
from warp.jax_experimental.ffi import jax_ad_kernel

jax.distributed.initialize()


# Kernels
@wp.kernel
def multiply_by_two_kernel(a_in: wp.array(dtype=wp.float32), a_out: wp.array(dtype=wp.float32)):
    index = wp.tid()
    a_out[index] = a_in[index] * 2.0


@wp.kernel
def scale_sum_square_kernel(
    a: wp.array(dtype=wp.float32), b: wp.array(dtype=wp.float32), s: float, out: wp.array(dtype=wp.float32)
):
    tid = wp.tid()
    out[tid] = (a[tid] * s + b[tid]) ** 2.0


# multi-output kernel
@wp.kernel
def multi_output_kernel(
    a: wp.array(dtype=wp.float32),
    b: wp.array(dtype=wp.float32),
    s: float,
    c: wp.array(dtype=wp.float32),
    d: wp.array(dtype=wp.float32),
):
    tid = wp.tid()
    c[tid] = a[tid] ** 2.0
    d[tid] = a[tid] * b[tid] * s


# JAX callables with AD
jax_mul = jax_ad_kernel(multiply_by_two_kernel, num_outputs=1)
jax_scale = jax_ad_kernel(scale_sum_square_kernel, num_outputs=1, static_argnames=("s",))
jax_mo = jax_ad_kernel(multi_output_kernel, num_outputs=2, static_argnames=("s",))


def warp_multiply(x):
    (out,) = jax_mul(x)
    return out


def warp_scale_sum_square(a, b, s):
    (out,) = jax_scale(a, b, s)
    return out


def warp_multi_output(a, b, s):
    c, d = jax_mo(a, b, s)
    return c, d


def make_sharded_1d(mesh, n):
    sharding = jax.sharding.NamedSharding(mesh, P("x"))
    a = jnp.arange(n, dtype=jnp.float32)
    shape = (n,)
    arrays = [jax.device_put(a[idx], d) for d, idx in sharding.addressable_devices_indices_map(shape).items()]
    return jax.make_array_from_single_device_arrays(shape, sharding, arrays)


def example_mul2_backward(mesh):
    n = jax.device_count() * 5
    x = make_sharded_1d(mesh, n)

    def loss(x):
        y = shard_map(lambda v: warp_multiply(v), mesh=mesh, in_specs=(P("x"),), out_specs=P("x"), check_rep=False)(x)
        return jnp.sum(y)

    g = jax.grad(loss)(x)
    gf = allgather(g)
    if jax.process_index() == 0:
        ref = np.full(n, 2.0, dtype=np.float32)
        np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
        print("mul2: OK")


def example_scale_sum_square_backward(mesh):
    n = jax.device_count() * 6
    a = make_sharded_1d(mesh, n)
    b = make_sharded_1d(mesh, n)
    s = 1.5

    def loss(a, b):
        y = shard_map(
            lambda aa, bb: warp_scale_sum_square(aa, bb, s),
            mesh=mesh,
            in_specs=(P("x"), P("x")),
            out_specs=P("x"),
            check_rep=False,
        )(a, b)
        return jnp.sum(y)

    da, db = jax.grad(loss, argnums=(0, 1))(a, b)
    daf, dbf = allgather(da), allgather(db)
    if jax.process_index() == 0:
        a_np = np.arange(n, dtype=np.float32)
        b_np = np.arange(n, dtype=np.float32)
        ref_da = 2.0 * (a_np * s + b_np) * s
        ref_db = 2.0 * (a_np * s + b_np)
        np.testing.assert_allclose(np.asarray(daf), ref_da, rtol=1e-5, atol=1e-6)
        np.testing.assert_allclose(np.asarray(dbf), ref_db, rtol=1e-5, atol=1e-6)
        print("scale_sum_square: OK")


def example_mul2_with_psum(mesh):
    n = jax.device_count() * 4
    x = make_sharded_1d(mesh, n)

    def body(v):
        local = warp_multiply(v)
        red = jnp.sum(local)
        total = jax.lax.psum(red, "x")
        return local, total

    def loss(x):
        local, total = shard_map(body, mesh=mesh, in_specs=(P("x"),), out_specs=(P("x"), P()), check_rep=False)(x)
        return jnp.sum(local) + total

    g = jax.grad(loss)(x)
    gf = allgather(g)
    if jax.process_index() == 0:
        ref = np.full(n, 4.0, dtype=np.float32)
        np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
        print("mul2 + psum: OK")


def example_psum_mean(mesh):
    n = jax.device_count() * 3
    x = make_sharded_1d(mesh, n)

    def body(v):
        local = warp_multiply(v)
        red = jnp.sum(local)
        total = jax.lax.psum(red, "x")
        mean = total / n
        return local, mean

    def loss(x):
        local, mean = shard_map(body, mesh=mesh, in_specs=(P("x"),), out_specs=(P("x"), P()), check_rep=False)(x)
        return jnp.sum(local) + mean

    g = jax.grad(loss)(x)
    gf = allgather(g)
    if jax.process_index() == 0:
        # d/dx sum(2x) = 2; d/dx mean(2x) = 2/n per element
        ref = np.full(n, 2.0 + 2.0 / n, dtype=np.float32)
        np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
        print("mul2 + psum mean: OK")


def example_multi_output_with_psum(mesh):
    n = jax.device_count() * 5
    a = make_sharded_1d(mesh, n)
    b = make_sharded_1d(mesh, n)
    s = 2.0

    def body(aa, bb):
        c, d = warp_multi_output(aa, bb, s)
        local = c + d
        total = jax.lax.psum(jnp.sum(local), "x")
        return local, total

    def loss(a, b):
        local, total = shard_map(body, mesh=mesh, in_specs=(P("x"), P("x")), out_specs=(P("x"), P()), check_rep=False)(
            a, b
        )
        return jnp.sum(local) + total

    da, db = jax.grad(loss, argnums=(0, 1))(a, b)
    daf, dbf = allgather(da), allgather(db)
    if jax.process_index() == 0:
        a_np = np.arange(n, dtype=np.float32)
        b_np = np.arange(n, dtype=np.float32)
        base_da = 2.0 * a_np + b_np * s
        base_db = a_np * s
        ref_da = 2.0 * base_da
        ref_db = 2.0 * base_db
        np.testing.assert_allclose(np.asarray(daf), ref_da, rtol=1e-5, atol=1e-6)
        np.testing.assert_allclose(np.asarray(dbf), ref_db, rtol=1e-5, atol=1e-6)
        print("multi_output + psum: OK")


def main():
    devices = jax.devices()
    mesh = jax.sharding.Mesh(np.array(devices), "x")
    example_mul2_backward(mesh)
    example_scale_sum_square_backward(mesh)
    example_mul2_with_psum(mesh)
    example_psum_mean(mesh)
    example_multi_output_with_psum(mesh)


if __name__ == "__main__":
    main()
`

You can run them with 
`mpirun -np 2 python test.py`

johnpjf commented:
Ok, is the intent to fix this multi-GPU issue before checking in?

On Tue, Aug 26, 2025 at 2:54 PM Mehdi Ataei ***@***.***>
wrote:

> *mehdiataei* left a comment (NVIDIA/warp#912)
> <https://github.com/NVIDIA/warp/pull/912#issuecomment-3225844742>
>
> Hey @johnpjf <https://github.com/johnpjf>
> Yes, if the examples are not executed with mpirun, we get that error. I’m
> not sure what the root cause is, but it’s probably a stream, event,
> function, or module created on GPU 0 being accidentally used while GPU 1 is
> current. Running with mpirun fixes it. I made the following test cases to
> check whether the grad works with shardmap, and they all pass successfully.
> However, I’m having trouble merging these tests with the current Warp test
> cases since they rely on mpirun.
>
> `import warp as wp
> import jax
> import jax.numpy as jnp
> import numpy as np
> from jax.sharding import PartitionSpec as P
> from jax.experimental.multihost_utils import process_allgather as allgather
> from jax.experimental.shard_map import shard_map
> from warp.jax_experimental.ffi import jax_ad_kernel
>
> jax.distributed.initialize()
> Kernels
>
> @wp.kernel
> def multiply_by_two_kernel(a_in: wp.array(dtype=wp.float32), a_out:
> wp.array(dtype=wp.float32)):
> index = wp.tid()
> a_out[index] = a_in[index] * 2.0
>
> @wp.kernel
> def scale_sum_square_kernel(
> a: wp.array(dtype=wp.float32), b: wp.array(dtype=wp.float32), s: float,
> out: wp.array(dtype=wp.float32)
> ):
> tid = wp.tid()
> out[tid] = (a[tid] * s + b[tid]) ** 2.0
> multi-output kernel
>
> @wp.kernel
> def multi_output_kernel(
> a: wp.array(dtype=wp.float32),
> b: wp.array(dtype=wp.float32),
> s: float,
> c: wp.array(dtype=wp.float32),
> d: wp.array(dtype=wp.float32),
> ):
> tid = wp.tid()
> c[tid] = a[tid] ** 2.0
> d[tid] = a[tid] * b[tid] * s
> JAX callables with AD
>
> jax_mul = jax_ad_kernel(multiply_by_two_kernel, num_outputs=1)
> jax_scale = jax_ad_kernel(scale_sum_square_kernel, num_outputs=1,
> static_argnames=("s",))
> jax_mo = jax_ad_kernel(multi_output_kernel, num_outputs=2,
> static_argnames=("s",))
>
> def warp_multiply(x):
> (out,) = jax_mul(x)
> return out
>
> def warp_scale_sum_square(a, b, s):
> (out,) = jax_scale(a, b, s)
> return out
>
> def warp_multi_output(a, b, s):
> c, d = jax_mo(a, b, s)
> return c, d
>
> def make_sharded_1d(mesh, n):
> sharding = jax.sharding.NamedSharding(mesh, P("x"))
> a = jnp.arange(n, dtype=jnp.float32)
> shape = (n,)
> arrays = [jax.device_put(a[idx], d) for d, idx in
> sharding.addressable_devices_indices_map(shape).items()]
> return jax.make_array_from_single_device_arrays(shape, sharding, arrays)
>
> def example_mul2_backward(mesh):
> n = jax.device_count() * 5
> x = make_sharded_1d(mesh, n)
>
> def loss(x):
>     y = shard_map(lambda v: warp_multiply(v), mesh=mesh, in_specs=(P("x"),), out_specs=P("x"), check_rep=False)(x)
>     return jnp.sum(y)
>
> g = jax.grad(loss)(x)
> gf = allgather(g)
> if jax.process_index() == 0:
>     ref = np.full(n, 2.0, dtype=np.float32)
>     np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
>     print("mul2: OK")
>
> def example_scale_sum_square_backward(mesh):
> n = jax.device_count() * 6
> a = make_sharded_1d(mesh, n)
> b = make_sharded_1d(mesh, n)
> s = 1.5
>
> def loss(a, b):
>     y = shard_map(
>         lambda aa, bb: warp_scale_sum_square(aa, bb, s),
>         mesh=mesh,
>         in_specs=(P("x"), P("x")),
>         out_specs=P("x"),
>         check_rep=False,
>     )(a, b)
>     return jnp.sum(y)
>
> da, db = jax.grad(loss, argnums=(0, 1))(a, b)
> daf, dbf = allgather(da), allgather(db)
> if jax.process_index() == 0:
>     a_np = np.arange(n, dtype=np.float32)
>     b_np = np.arange(n, dtype=np.float32)
>     ref_da = 2.0 * (a_np * s + b_np) * s
>     ref_db = 2.0 * (a_np * s + b_np)
>     np.testing.assert_allclose(np.asarray(daf), ref_da, rtol=1e-5, atol=1e-6)
>     np.testing.assert_allclose(np.asarray(dbf), ref_db, rtol=1e-5, atol=1e-6)
>     print("scale_sum_square: OK")
>
> def example_mul2_with_psum(mesh):
> n = jax.device_count() * 4
> x = make_sharded_1d(mesh, n)
>
> def body(v):
>     local = warp_multiply(v)
>     red = jnp.sum(local)
>     total = jax.lax.psum(red, "x")
>     return local, total
>
> def loss(x):
>     local, total = shard_map(body, mesh=mesh, in_specs=(P("x"),), out_specs=(P("x"), P()), check_rep=False)(x)
>     return jnp.sum(local) + total
>
> g = jax.grad(loss)(x)
> gf = allgather(g)
> if jax.process_index() == 0:
>     ref = np.full(n, 4.0, dtype=np.float32)
>     np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
>     print("mul2 + psum: OK")
>
> def example_psum_mean(mesh):
> n = jax.device_count() * 3
> x = make_sharded_1d(mesh, n)
>
> def body(v):
>     local = warp_multiply(v)
>     red = jnp.sum(local)
>     total = jax.lax.psum(red, "x")
>     mean = total / n
>     return local, mean
>
> def loss(x):
>     local, mean = shard_map(body, mesh=mesh, in_specs=(P("x"),), out_specs=(P("x"), P()), check_rep=False)(x)
>     return jnp.sum(local) + mean
>
> g = jax.grad(loss)(x)
> gf = allgather(g)
> if jax.process_index() == 0:
>     # d/dx sum(2x) = 2; d/dx mean(2x) = 2/n per element
>     ref = np.full(n, 2.0 + 2.0 / n, dtype=np.float32)
>     np.testing.assert_allclose(np.asarray(gf), ref, rtol=1e-5, atol=1e-6)
>     print("mul2 + psum mean: OK")
>
> def example_multi_output_with_psum(mesh):
> n = jax.device_count() * 5
> a = make_sharded_1d(mesh, n)
> b = make_sharded_1d(mesh, n)
> s = 2.0
>
> def body(aa, bb):
>     c, d = warp_multi_output(aa, bb, s)
>     local = c + d
>     total = jax.lax.psum(jnp.sum(local), "x")
>     return local, total
>
> def loss(a, b):
>     local, total = shard_map(body, mesh=mesh, in_specs=(P("x"), P("x")), out_specs=(P("x"), P()), check_rep=False)(
>         a, b
>     )
>     return jnp.sum(local) + total
>
> da, db = jax.grad(loss, argnums=(0, 1))(a, b)
> daf, dbf = allgather(da), allgather(db)
> if jax.process_index() == 0:
>     a_np = np.arange(n, dtype=np.float32)
>     b_np = np.arange(n, dtype=np.float32)
>     base_da = 2.0 * a_np + b_np * s
>     base_db = a_np * s
>     ref_da = 2.0 * base_da
>     ref_db = 2.0 * base_db
>     np.testing.assert_allclose(np.asarray(daf), ref_da, rtol=1e-5, atol=1e-6)
>     np.testing.assert_allclose(np.asarray(dbf), ref_db, rtol=1e-5, atol=1e-6)
>     print("multi_output + psum: OK")
>
> def main():
> devices = jax.devices()
> mesh = jax.sharding.Mesh(np.array(devices), "x")
> example_mul2_backward(mesh)
> example_scale_sum_square_backward(mesh)
> example_mul2_with_psum(mesh)
> example_psum_mean(mesh)
> example_multi_output_with_psum(mesh)
>
> if *name* == "*main*":
> main()
> `
>
> You can run them with
> mpirun -np 2 python test.py
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/warp/pull/912#issuecomment-3225844742>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHLB5C5Q5BAELOBHTAPMFST3PTJPXAVCNFSM6AAAAACD5LNHJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMRVHA2DINZUGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mehdiataei commented:
It is fairly common to run multi-GPU workloads with multiple processes in JAX even on a single machine, so I don’t think this is a blocking issue (and this limitation existed before this PR). I have some ideas about the root cause and a potential fix, so let me take a shot at it first.

johnpjf commented:
Yes, multi-GPU training within a pmap is essentially a requirement for all of our workloads so it would be great to get a fix for this. Thanks

mehdiataei commented:
@johnpjf Good news. I was able to resolve the CUDA 400 issue by launching FFI on callframe CUDA stream and preload modules across GPUs to prevent per-device race conditions. With these changes in place, your example now runs correctly without requiring the use of `mpirun`. Now any multi-GPU (and distributed multi-GPU with MPI) should be working similar to JAX.

This fix also made it possible to integrate both the `shardmap` and `pmap` test cases directly into `test_jax.py`. 

Can you give it a try? Thanks.

mehdiataei commented:
Hmm I can't reproduce it:
<img width="688" height="925" alt="image" src="https://github.com/user-attachments/assets/de64be2b-95b6-4f27-8e6c-3b400764aa7c" />

Can you delete the kernel cache folder and try again?

johnpjf commented:
Tried that, but didn't fix it. Looks like we have a copy of warp in our
monorepo that's from 8/7, is it possible that your change depends on
something since then. I'll try to get ours updated and try again tomorrow.
Thanks again for working on this!

On Tue, Aug 26, 2025 at 4:56 PM Mehdi Ataei ***@***.***>
wrote:

> *mehdiataei* left a comment (NVIDIA/warp#912)
> <https://github.com/NVIDIA/warp/pull/912#issuecomment-3226151688>
>
> Hmm I can't reproduce it:
> image.png (view on web)
> <https://github.com/user-attachments/assets/de64be2b-95b6-4f27-8e6c-3b400764aa7c>
>
> Can you delete the kernel cache folder and try again?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NVIDIA/warp/pull/912#issuecomment-3226151688>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHLB5C6YUW7SS4NDXANYS4D3PTX2RAVCNFSM6AAAAACD5LNHJWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMRWGE2TCNRYHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

johnpjf commented:
Ok, turns out I have a bad GPU, pmap works great on a different dual-GPU machine, thanks!

mehdiataei commented:
Great! I think for JAX to work you need to have identical GPUs (maybe that's the cause)

nouiz commented:
> Great! I think for JAX to work you need to have identical GPUs (maybe that's the cause)

Exact. You can't mix different kinds of GPUs. This isn't supported.

btaba commented:
> @btaba could you share more details on the [WIP](https://github.com/google-deepmind/mujoco/compare/main...btaba:mujoco:btaba-patch-1?expand=1) ? I played around with this a bit over the weekend and it looks like the issue comes from how modules are being loaded in the “third_party” FFI (in Mujoco). I made some progress but was planning to dig deeper into it this week. Is this the issue you’re working on fixing?

I'm not actively working on it, just wanted to take a stab to try to get to a semi-working state. If you have some cycles, please let us know what you find!

johnpjf commented on warp/jax_experimental/ffi.py:
```diff
@@ -823,6 +905,292 @@ def ffi_callback(call_frame):
+    Args:
+        kernel: The Warp kernel to wrap.
+        num_outputs: Number of output arrays produced by the kernel.
+        static_argnames: Optional iterable of argument names that should be treated
```
It seems there is some confusion in this function between static-for-jit arguments and arguments that you don't want the gradients for. I can have a jax array as an input that is not static-for-jit but that I don't want gradients for. For example, when ray tracing, I don't want gradients of rays, but I don't want them to be static for jit (i.e. they are regular JAX Arrays).

johnpjf replied:
Actually I think my comment was a little wrong, but I do think this code slightly conflates static args for jitting and for custom gradients. I don't really think this code should be concerned with jit-ing at all, static args for this might not be (and usually won't be!) static args for jit.
For specific changes:
Also, you should call these `nondiff_argnums` to match JAX's terminology. https://docs.jax.dev/en/latest/_autosummary/jax.custom_jvp.html
I also wouldn't try to be clever and use heuristics to figure what should or should not be `nondiff_argnums`, that could cause silent bugs where someone forgets to wrap a float in a `jnp.array`.
I think we should also allow for JAX input array's that shouldn't have gradients computed (i.e. as if we had wrapped them in `jax.stop_gradient`). I don't know what to call these or if Warp really supports *not* computing gradients correctly.

johnpjf commented on warp/jax_experimental/ffi.py:
```diff
@@ -823,6 +905,292 @@ def ffi_callback(call_frame):
+            as static (non-differentiable) in JAX. If None, scalar (non-array) inputs
+            are treated as static by default.
+        vmap_method: How the callback transforms under jax.vmap.
+        launch_dim_arg_index: Index (in the kernel's argument list) of the input
```
(Sorry was in wrong place before)
I don't this will work for everyone, for example if I want to do something for every pixel in an RGB image, then I want the kernel to be over the height,width but not channel dimension. Also, this is different than the existing jax_kernel interface. I wonder if it's better to force providing these explicitly?

mehdiataei commented on warp/jax_experimental/ffi.py:
```diff
@@ -823,6 +905,292 @@ def ffi_callback(call_frame):
+
+    # If static scalar args are present, wrap with jitted callable that marks them static
+    if static_args:
+        static_names = [parameters[i].name for i in static_args]
```
For static argnums to be registered we need to call jit
https://docs.jax.dev/en/latest/_autosummary/jax.jit.html
This is only used if the user provide the static arguments (or if there are static inputs such as primitive types)

btaba commented:
> > What am I missing?
> 
> I digged into how Mujoco uses Warp. It seems like it uses jax_callable function instead of jax_(ad)_kernel. I went ahead and added adjoint support for jax_callable as well (as the first step to fix all the issue). BUT it seems like there is still issues with grabbing the exact device of the CUDA stream that XLA gives us when using pmap. If you run the pmap tests, sometime they pass sometime they don't!! Unfortunately the XLA Python FFI doesn’t expose the device ordinal; we need a tiny native helper in Warp to query it from the stream/context (I did add this wrapper in C++ but still haven't been able to fix this issue). Do you know of any way to get teh device ordinal?
> 
> Another option is to use the Warp kernels as kernels (using jax_ad_kernel) which mitigates this issue all together, it shouldn't require much change on the mujoco side (and it works perfectly well under pmap).
> 
> btw, when testing, pls set `wp.config.enable_backward = True` right after importing mjx, because the constraint.py turns it off. Otherwise you see another error if you use jax_ad_kernel with mjx.

Thanks for taking a look! This is why I had the hack around device pointers, to get `device_ordinal` directly from the buffer. I'm not sure of a better way. Perhaps someone on the NVIDIA side can help here, in which case, our lives would be made much easier. 

What in jax_ad_kernel makes it work with pmap, but a regular warp kernel does not? Is there a way to tease apart the pmap support from the grad support? That would be really amazing to get pmap support into the main branch soon-ish. 

@nvlukasz for visibility

mehdiataei commented:
> More generally, can jax_ad_kernel just replace jax_kernel.

Yes. One is enough. I haven't removed it yet but will do after @nvlukasz review and confirm.

mehdiataei commented:
> > > What am I missing?
> > 
> > 
> > I digged into how Mujoco uses Warp. It seems like it uses jax_callable function instead of jax_(ad)_kernel. I went ahead and added adjoint support for jax_callable as well (as the first step to fix all the issue). BUT it seems like there is still issues with grabbing the exact device of the CUDA stream that XLA gives us when using pmap. If you run the pmap tests, sometime they pass sometime they don't!! Unfortunately the XLA Python FFI doesn’t expose the device ordinal; we need a tiny native helper in Warp to query it from the stream/context (I did add this wrapper in C++ but still haven't been able to fix this issue). Do you know of any way to get teh device ordinal?
> > Another option is to use the Warp kernels as kernels (using jax_ad_kernel) which mitigates this issue all together, it shouldn't require much change on the mujoco side (and it works perfectly well under pmap).
> > btw, when testing, pls set `wp.config.enable_backward = True` right after importing mjx, because the constraint.py turns it off. Otherwise you see another error if you use jax_ad_kernel with mjx.
> 
> Thanks for taking a look! This is why I had the hack around device pointers, to get `device_ordinal` directly from the buffer. I'm not sure of a better way. Perhaps someone on the NVIDIA side can help here, in which case, our lives would be made much easier.
> 
> What in jax_ad_kernel makes it work with pmap, but the jax_callable does not? Is there a way to tease apart the pmap support from the grad support? That would be really amazing to get pmap support into the main branch soon-ish.
> 
> @nvlukasz for visibility

Thanks. The jax_ad_kernel simply launches the adjoint kernels created by Warp, whearas in jax_callable we need to use Tape to record the order of computations (because jax_callable is a generic wrapper for arbitrary annotated Python functions that may contain multiple Warp ops and not a single kernel with a compiled adjoint). 

Removing the adjoint part really won't make any difference for the current issue we have for the pmap support. We'll have the exact same issue for forward as well. If we fix it the adjoint computation will just work without any issues. 

The solution is probably on the C++/CUDA world. I am looking into how to get the device ordinal. There is hope.

nvlukasz commented:
> The solution is probably on the C++/CUDA world. I am looking into how to get the device ordinal. There is hope.

So we can't get the CUDA device from JAX, via `get_jax_device()` or similar?

I suppose it's not possible to wrap code in Python contexts like `jax.default_device` or `wp.ScopedDevice`?

I don't see a way to get the device from XLA, but we can get the stream. Then we can use `cuStreamGetCtx()` under the hood to figure out the device. IIRC, one caveat with that function is that it's not allowed during graph capture, so I've been avoiding it in general. We'd need to be careful there.

We can also query the current CUDA context in Warp, we have some functions for that already. But if JAX does not set it, then this approach won't work (same with using the special device `"cuda"`).

Another approach is to get the device from the JAX arrays or buffers, and assume that the launch is to take place on the same device. There is a way to query the CUDA context associated with a pointer (though I think it has the same caveats with graph capture as `cuStreamGetCtx`).

If none of these work, I'm slightly at a loss. Maybe it's a question for JAX or XLA folks?

mehdiataei commented:
Ok… finally I was able to fix the device ordinal match by adding new C++ APIs to Warp (it wasn't an easy fix, it took a bit of trial and error). `jax_callable` now fully supports both `pmap` for forward and adjoint, and I tested them extensively (many test cases are added). I did the tests in a long loop as well to avoid false positives due to randomness.

I am now working on fixing this on the Mujoco side, but the graph capture in Mujoco is a bit convoluted and relies heavily on various capture statements, which I think are causing issues.

P.S. Although I don’t think removing adjoint will help fixing the current issue, disabling it is straightforward: just avoid passing the backward function and set auto_grad to false for jax_callable (which is the only thing used on mujoco side). If reviewing is simpler with it separated, we can do that, but I don’t think it will help with the problem.

mehdiataei commented:
Good news! I think I fixed mojuco issue as well (Seems like the backward is not enabled for some of the functions which caused some issues). Summary:

Fixed two issues that broke pmap+vmap with MJX Warp FFI:
  1) Removed Warp CUDA graph capture inside the FFI callback when running on the XLA stream.
  2) Temporarily disabled Warp backward codegen during the FFI callback.

this helps because the original FFI wrapper tried to capture a CUDA graph on the XLA stream and also called Warp APIs that can insert conditional nodes (`wp.capture_if`). CUDA’s capture is thread-local and stream-bound, so mixing XLA’s external stream capture with Warp’s internal capture paths caused “Graph capture is not active on this stream” and related insert/pause errors under parallel execution. Separately, Warp was JIT-building adjoint kernels (enable_backward=True) for collision/narrowphase code during the callback. Under pmap/vmap and without a stable capture context, that triggered NVRTC compile issues and module load mismatches

So by wrapping the call with a small context that sets `wp.config.enable_backward = False` for the duration of the callback, guarded by a lock, I prevent on-the-fly generation of backward kernels while still allowing forward kernels to compile and run. Non-XLA stream path retains Warp’s optional graph capture.

Notes
  - All adjoint computations are unaffected and all tests pass becuase already built kernels (including their adjoints) are unaffected.

  ```
  ----------------------------------------------------------------------
Ran 64 tests in 120.365s

OK
```
  - If/when gradients are needed through these kernels, we can revisit enabling backward selectively or prebuilding the required adjoint kernels outside the callback so they don’t compile during pmap/vmap execution. If a kernel is first-time compiled while inside the callback, it will be compiled without an adjoint. If you later want gradients through that specific kernel via Warp’s Tape, you’ll need to prebuild it with backward enabled (or clear cache and rebuild) outside the callback.

To test yourself copy the content of `ffi.py` to the mujoco and test with the following:

**IMPORTANT: make sure to rebuild Warp from source since changes were made in the C++ code.**

@btaba example:

```python
import mujoco
from mujoco import mjx
import jax


xml = """
<mujoco>
  <worldbody>
    <light name="top" pos="0 0 1"/>
    <body name="floor" pos="0 0 0">
      <geom name="floor" type="plane" size="1 1 0.1" rgba=".8 .8 .8 1"/>
    </body>
    <body name="sphere" pos="0 0 1">
      <joint type="free"/>
      <geom name="sphere" type="sphere" size="0.1" rgba="0 0 1 1"/>
    </body>
  </worldbody>
</mujoco>
"""

m = mujoco.MjModel.from_xml_string(xml)
mx = mjx.put_model(m, impl='warp')

def make_data(rng):
  dx = mjx.make_data(m, impl='warp', nconmax=100, njmax=10)
  dx = dx.replace(qpos=jax.random.uniform(rng, mx.nq))
  return dx

device_count = jax.local_device_count()
assert device_count > 1
batch_size = 16

per_device_keys = jax.random.split(jax.random.PRNGKey(0), device_count)
dx_batch = jax.pmap(lambda k: jax.vmap(make_data)(jax.random.split(k, batch_size)))(per_device_keys)
assert dx_batch.qpos.shape == (device_count, batch_size, m.nq)

out = jax.pmap(jax.vmap(mjx.step, in_axes=(None, 0)), in_axes=(None, 0))(mx, dx_batch)
print(out.time)
assert (out.time == 0.002).all()
```

output:
```
Warp 1.8.1 initialized:
   CUDA Toolkit 12.9, Driver 12.9
   Devices:
     "cpu"      : "x86_64"
     "cuda:0"   : "NVIDIA RTX 6000 Ada Generation" (47 GiB, sm_89, mempool enabled)
     "cuda:1"   : "NVIDIA RTX 6000 Ada Generation" (47 GiB, sm_89, mempool enabled)
   CUDA peer access:
     Supported fully (all-directional)
   Kernel cache:
     /home/mehdi/.cache/warp/1.8.1
Module mujoco.mjx.third_party.mujoco_warp._src.smooth 0c1050e load on device 'cuda:0' took 8.68 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_driver dc593c3 load on device 'cuda:0' took 0.30 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_driver._nxn_broadphase 2441629 load on device 'cuda:0' took 0.27 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_primitive 00c6f22 load on device 'cuda:0' took 0.31 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.constraint 7a9fd39 load on device 'cuda:0' took 1.39 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.forward 76a01ca load on device 'cuda:0' took 0.61 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.passive 4223120 load on device 'cuda:0' took 0.54 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.support aa34105 load on device 'cuda:0' took 0.34 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.smooth._tile_cholesky_factorize_solve a1700c8 load on device 'cuda:0' took 26.94 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver b37b8b0 load on device 'cuda:0' took 1.16 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.support.mul_m_dense 397bd0e load on device 'cuda:0' took 21.33 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.update_constraint_gauss_cost cd535c8 load on device 'cuda:0' took 0.26 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.update_gradient_cholesky 4039252 load on device 'cuda:0' took 28.84 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.linesearch_jv_fused 37529ee load on device 'cuda:0' took 0.31 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.forward._tile_euler_dense 756e4cd load on device 'cuda:0' took 29.54 ms  (cached)
Module mujoco.mjx.warp.ffi c043961 load on device 'cuda:0' took 0.28 ms  (cached)
Module mujoco.mjx.warp.ffi c043961 load on device 'cuda:1' took 0.28 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.smooth 0c1050e load on device 'cuda:1' took 13.61 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_driver dc593c3 load on device 'cuda:1' took 0.65 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_driver._nxn_broadphase 2441629 load on device 'cuda:1' took 0.61 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.collision_primitive 00c6f22 load on device 'cuda:1' took 0.31 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.constraint 7a9fd39 load on device 'cuda:1' took 1.56 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.forward 76a01ca load on device 'cuda:1' took 0.72 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.passive 4223120 load on device 'cuda:1' took 0.53 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.support aa34105 load on device 'cuda:1' took 0.33 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.smooth._tile_cholesky_factorize_solve a1700c8 load on device 'cuda:1' took 29.98 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver b37b8b0 load on device 'cuda:1' took 1.15 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.support.mul_m_dense 397bd0e load on device 'cuda:1' took 22.49 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.update_constraint_gauss_cost cd535c8 load on device 'cuda:1' took 0.24 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.update_gradient_cholesky 4039252 load on device 'cuda:1' took 29.70 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.solver.linesearch_jv_fused 37529ee load on device 'cuda:1' took 0.33 ms  (cached)
Module mujoco.mjx.third_party.mujoco_warp._src.forward._tile_euler_dense 756e4cd load on device 'cuda:1' took 31.55 ms  (cached)
[[0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002
  0.002 0.002 0.002 0.002]
 [0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002 0.002
  0.002 0.002 0.002 0.002]]
 ```

Enjoy!!

nvlukasz commented:
> > I don't see a way to get the device from XLA.
> 
> There is a way actually. We should be able to use `XLA_FFI_DeviceOrdinal_Get` from here: https://github.com/openxla/xla/blob/97b0e76e6c706fe0a7b22e02a9d8e6949b417ee9/xla/ffi/ffi_api.cc#L632
> 
> We would need to add it the XLA_FFI_Api list [here](https://github.com/NVIDIA/warp/blob/46dc8b6a17ae9ec381e8a133420e1bba8904ec00/warp/jax_experimental/xla_ffi.py#L406) and plumb it through. That might be safer/easier than trying to get it from the stream or array pointers.

Ah, that sounds like just what the doctor ordered. Thanks for pointing it out, @chaserileyroberts! @mehdiataei FYI

mehdiataei commented:
Thanks @chaserileyroberts . We can definitely switch to that when the API is exposed. Until then the pointer monkey patch should be workable.

btaba commented:
Thanks @mehdiataei , I just tried with the latest branch but still get zeros on the second device. I'll have to try on another machine; do you have any ideas otherwise?

nvlukasz commented:
> Thanks @mehdiataei , I just tried with the latest branch but still get zeros on the second device. I'll have to try on another machine; do you have any ideas otherwise?

@btaba @mehdiataei, I have a pmap solution based on [GH-976](https://github.com/NVIDIA/warp/pull/976) from @chaserileyroberts. Thanks Chase! It's currently under review by the Warp team, but I expect it to be merged shortly.

I had to add some additional thread safety, but the Mujoco example works fine now. That should take care of pmap, so once that's merged, @mehdiataei you can modify this PR to only add the autodiff functionality.

Thank you all for your contributions and patience!

nvlukasz commented:
> @nvlukasz That's great! Yes. I can merge those changes and add the adjoint stuff.

Alright, the new pmap support is now merged (https://github.com/NVIDIA/warp/commit/599ee98c9c1f4be1f37a18279335f3ca954b1d48)

shi-eric commented on warp/examples/optim/example_inverse_kinematics_jax.py:
This example cannot be added since `warp.sim` is now removed from the `main` branch. Maybe it can be contributed to `newton` instead.
